From 13b5daa777e9ffea85c0b9bb92c6145920a6ced5 Mon Sep 17 00:00:00 2001
From: Tobias Kaufmann <Tobias.KA.Kaufmann@bmw.de>
Date: Tue, 13 Dec 2022 14:46:35 +0100
Subject: [PATCH 1/8] CMakeLists.txt: Add BUILD_TESTING compile flag

Up to now all tests were built automatically when building with cmake.

As tests are not required when just building the library a new cmake
flag BUILD_TESTING is introduced. Once set the test binaries are build
and can be executed using ctest.
---
 CMakeLists.txt     |   7 ++-
 ref/CMakeLists.txt | 132 +++++++++++++++++++++++++++++----------------
 2 files changed, 90 insertions(+), 49 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 17b9f75..60cb20e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,10 +5,13 @@ project(dilithium C ASM)
 set(CMAKE_C_STANDARD 99)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 
-enable_testing()
-
 #find_package(OpenSSL REQUIRED)
 #include_directories(${OPENSSL_INCLUDE_DIR})
 
+if(BUILD_TESTING)
+    message(STATUS "Testing enabled.")
+    enable_testing()
+endif()
+
 add_subdirectory(ref)
 #add_subdirectory(avx2)
diff --git a/ref/CMakeLists.txt b/ref/CMakeLists.txt
index b6f76e0..c246b1e 100644
--- a/ref/CMakeLists.txt
+++ b/ref/CMakeLists.txt
@@ -26,14 +26,20 @@ target_compile_definitions(dilithium2aes_ref PUBLIC DILITHIUM_MODE=2 DILITHIUM_U
 target_link_libraries(dilithium2_ref INTERFACE fips202_ref)
 target_link_libraries(dilithium2aes_ref INTERFACE fips202_ref aes256ctr_ref)
 
-add_executable(test_dilithium2_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors2_ref ${TEST_VECTORS_SRCS})
-add_executable(test_dilithium2aes_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors2aes_ref ${TEST_VECTORS_SRCS})
-target_link_libraries(test_dilithium2_ref dilithium2_ref)
-target_link_libraries(test_vectors2_ref dilithium2_ref)
-target_link_libraries(test_dilithium2aes_ref dilithium2aes_ref)
-target_link_libraries(test_vectors2aes_ref dilithium2aes_ref)
+if(BUILD_TESTING)
+  add_executable(test_dilithium2_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors2_ref ${TEST_VECTORS_SRCS})
+  add_executable(test_dilithium2aes_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors2aes_ref ${TEST_VECTORS_SRCS})
+
+  target_link_libraries(test_dilithium2_ref dilithium2_ref)
+  target_link_libraries(test_vectors2_ref dilithium2_ref)
+  target_link_libraries(test_dilithium2aes_ref dilithium2aes_ref)
+  target_link_libraries(test_vectors2aes_ref dilithium2aes_ref)
+
+  add_test(NAME dilithium2_ref COMMAND test_dilithium2_ref)
+  add_test(NAME dilithium2aes_ref COMMAND test_dilithium2aes_ref)
+endif()
 
 # Dilithium 3
 add_library(dilithium3_ref ${DILITHIUM_FIPS202_SRCS})
@@ -43,14 +49,18 @@ target_compile_definitions(dilithium3aes_ref PUBLIC DILITHIUM_MODE=3 DILITHIUM_U
 target_link_libraries(dilithium3_ref INTERFACE fips202_ref)
 target_link_libraries(dilithium3aes_ref INTERFACE fips202_ref aes256ctr_ref)
 
-add_executable(test_dilithium3_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors3_ref ${TEST_VECTORS_SRCS})
-add_executable(test_dilithium3aes_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors3aes_ref ${TEST_VECTORS_SRCS})
-target_link_libraries(test_dilithium3_ref dilithium3_ref)
-target_link_libraries(test_vectors3_ref dilithium3_ref)
-target_link_libraries(test_dilithium3aes_ref dilithium3aes_ref)
-target_link_libraries(test_vectors3aes_ref dilithium3aes_ref)
+if(BUILD_TESTING)
+  add_executable(test_dilithium3_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors3_ref ${TEST_VECTORS_SRCS})
+  add_executable(test_dilithium3aes_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors3aes_ref ${TEST_VECTORS_SRCS})
+  target_link_libraries(test_dilithium3_ref dilithium3_ref)
+  target_link_libraries(test_vectors3_ref dilithium3_ref)
+  target_link_libraries(test_dilithium3aes_ref dilithium3aes_ref)
+  target_link_libraries(test_vectors3aes_ref dilithium3aes_ref)
+  add_test(NAME dilithium3_ref COMMAND test_dilithium3_ref)
+  add_test(NAME dilithium3aes_ref COMMAND test_dilithium3aes_ref)
+endif()
 
 # Dilithium 5
 add_library(dilithium5_ref ${DILITHIUM_FIPS202_SRCS})
@@ -60,36 +70,64 @@ target_compile_definitions(dilithium5aes_ref PUBLIC DILITHIUM_MODE=5 DILITHIUM_U
 target_link_libraries(dilithium5_ref INTERFACE fips202_ref)
 target_link_libraries(dilithium5aes_ref INTERFACE fips202_ref aes256ctr_ref)
 
-add_executable(test_dilithium5_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors5_ref ${TEST_VECTORS_SRCS})
-add_executable(test_dilithium5aes_ref ${TEST_DILITHIUM_SRCS})
-add_executable(test_vectors5aes_ref ${TEST_VECTORS_SRCS})
-target_link_libraries(test_dilithium5_ref dilithium5_ref)
-target_link_libraries(test_vectors5_ref dilithium5_ref)
-target_link_libraries(test_dilithium5aes_ref dilithium5aes_ref)
-target_link_libraries(test_vectors5aes_ref dilithium5aes_ref)
-
-add_test(NAME dilithium2_ref COMMAND test_dilithium2_ref)
-add_test(NAME dilithium2aes_ref COMMAND test_dilithium2aes_ref)
-add_test(NAME dilithium3_ref COMMAND test_dilithium3_ref)
-add_test(NAME dilithium3aes_ref COMMAND test_dilithium3aes_ref)
-add_test(NAME dilithium5_ref COMMAND test_dilithium5_ref)
-add_test(NAME dilithium5aes_ref COMMAND test_dilithium5aes_ref)
-
-if(WIN32)
-  add_test(NAME vectors2_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors2_ref> | dos2unix > tvecs2")
-  add_test(NAME vectors2aes_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors2aes_ref> | dos2unix > tvecs2aes")
-  add_test(NAME vectors3_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors3_ref> | dos2unix > tvecs3")
-  add_test(NAME vectors3aes_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors3aes_ref> | dos2unix > tvecs3aes")
-  add_test(NAME vectors5_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors5_ref> | dos2unix > tvecs5")
-  add_test(NAME vectors5aes_ref COMMAND PowerShell -Command "$<TARGET_FILE:test_vectors5aes_ref> | dos2unix > tvecs5aes")
-else()
-  add_test(NAME vectors2_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors2_ref>\" > tvecs2")
-  add_test(NAME vectors2aes_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors2aes_ref>\" > tvecs2aes")
-  add_test(NAME vectors3_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors3_ref>\" > tvecs3")
-  add_test(NAME vectors3aes_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors3aes_ref>\" > tvecs3aes")
-  add_test(NAME vectors5_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors5_ref>\" > tvecs5")
-  add_test(NAME vectors5aes_ref COMMAND sh -c "\"$<TARGET_FILE:test_vectors5aes_ref>\" > tvecs5aes")
+
+if(BUILD_TESTING)
+  add_executable(test_dilithium5_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors5_ref ${TEST_VECTORS_SRCS})
+  add_executable(test_dilithium5aes_ref ${TEST_DILITHIUM_SRCS})
+  add_executable(test_vectors5aes_ref ${TEST_VECTORS_SRCS})
+  target_link_libraries(test_dilithium5_ref dilithium5_ref)
+  target_link_libraries(test_vectors5_ref dilithium5_ref)
+  target_link_libraries(test_dilithium5aes_ref dilithium5aes_ref)
+  target_link_libraries(test_vectors5aes_ref dilithium5aes_ref)
+
+  add_test(NAME dilithium5_ref COMMAND test_dilithium5_ref)
+  add_test(NAME dilithium5aes_ref COMMAND test_dilithium5aes_ref)
+endif()
+
+
+# Vector Tests
+if(BUILD_TESTING)
+  if(WIN32)
+    add_test(NAME vectors2_ref
+             COMMAND PowerShell -Command
+                     "$<TARGET_FILE:test_vectors2_ref> | dos2unix > tvecs2")
+    add_test(
+      NAME vectors2aes_ref
+      COMMAND PowerShell -Command
+              "$<TARGET_FILE:test_vectors2aes_ref> | dos2unix > tvecs2aes")
+    add_test(NAME vectors3_ref
+             COMMAND PowerShell -Command
+                     "$<TARGET_FILE:test_vectors3_ref> | dos2unix > tvecs3")
+    add_test(
+      NAME vectors3aes_ref
+      COMMAND PowerShell -Command
+              "$<TARGET_FILE:test_vectors3aes_ref> | dos2unix > tvecs3aes")
+    add_test(NAME vectors5_ref
+             COMMAND PowerShell -Command
+                     "$<TARGET_FILE:test_vectors5_ref> | dos2unix > tvecs5")
+    add_test(
+      NAME vectors5aes_ref
+      COMMAND PowerShell -Command
+              "$<TARGET_FILE:test_vectors5aes_ref> | dos2unix > tvecs5aes")
+  else()
+    add_test(NAME vectors2_ref
+             COMMAND sh -c "\"$<TARGET_FILE:test_vectors2_ref>\" > tvecs2")
+    add_test(NAME vectors2aes_ref
+             COMMAND sh -c
+                     "\"$<TARGET_FILE:test_vectors2aes_ref>\" > tvecs2aes")
+    add_test(NAME vectors3_ref
+             COMMAND sh -c "\"$<TARGET_FILE:test_vectors3_ref>\" > tvecs3")
+    add_test(NAME vectors3aes_ref
+             COMMAND sh -c
+                     "\"$<TARGET_FILE:test_vectors3aes_ref>\" > tvecs3aes")
+    add_test(NAME vectors5_ref
+             COMMAND sh -c "\"$<TARGET_FILE:test_vectors5_ref>\" > tvecs5")
+    add_test(NAME vectors5aes_ref
+             COMMAND sh -c
+                     "\"$<TARGET_FILE:test_vectors5aes_ref>\" > tvecs5aes")
+  endif()
+
+  add_test(NAME hashes COMMAND sha256sum -c ../../SHA256SUMS)
 endif()
 
-add_test(NAME hashes COMMAND sha256sum -c ../../SHA256SUMS)
-- 
2.38.1

